modules/percolator/src/main/java/org/elasticsearch/percolator/ExtractQueryTermsService.java

```diff
@@ -104,24 +118,30 @@ public static void extractQueryTerms(Query query, ParseContext.Document document
      * If from part of the query, no query terms can be extracted then term extraction is stopped and
      * an UnsupportedQueryException is thrown.
      */
-    static Set<Term> extractQueryTerms(Query query) {
-        if (query instanceof MatchNoDocsQuery) {
-            // no terms to extract as this query matches no docs
-            return Collections.emptySet();
+    // TODO: restructure this method? For example have a Function<Query, Result> per supported query.
+    // All functions then sit in a Map<Class<? extends Query>, Function<Query, Result>> and we lookup via map.get(query.getClass())
+    static Result extractQueryTerms(Query query) {
+        if (query instanceof MatchAllDocsQuery) {
+            // we need to emit an empty term otherwise a percolator query with this query may never end up as a
+            // candidate match:
+            return new Result(true, new Term("", ""));
+        } else  if (query instanceof MatchNoDocsQuery) {
+            return new Result(false);
```

Comment:
It is true that extraction is complete, but we use the `verified` boolean also as an indication to skip memory index verification and emit if a doc is a match. In the case that a percolator query contains just a MatchNoDocsQuery (happens if `match` query text analysis doesn't emit any token) then this would kind of be an illegal indication.

Maybe complete should have two outcomes? `complete_match` and `complete_no_match`?